### PR TITLE
Ref/idw2

### PR DESCRIPTION
**Pull request by oed**
Refactors IDW to be more modular.

ThreeId provider is now also a DIDprovider.

Creates 3ID on creation.

### CODE REVIEW

File: src/three-idx.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  async loadStructure(accountId: string) {
+  }
+
+  async encodeKidWithVersion(keyName: string = 'signing'): Promise<string> {
+    const version = (await this.ceramic.listVersions(this.docs['3id'].id)).pop() || 0
+    return `${this.DID}?version=${version}#${keyName}`
+  }
+
+  parseKeyName(kid: string): string {
```
Comment (PaulLeCam) [inline src/three-idx.ts:33]: Seems the return type should be `string | null`

File: src/threeIdProvider.ts [hunk 2]
```diff
@@ -19,94 +20,98 @@ type Context = {
...
+  '3id_isAuthenticated': async ({ permissions, origin }, params) => {
+    return permissions.has(origin, params.spaces)
   },
-  '3id_signClaim': async (ctx, params) => {
-    return await ctx.provider.wallet.signClaim(params.payload, {
-      DID: params.did,
-      space: params.space,
+  '3id_signClaim': async ({ threeIdx, permissions, keyring, origin }, params) => {
+    if (!permissions.has(origin, params.spaces)) {
+      throw new RPCError(0, 'Authentication required')
```
Comment (PaulLeCam) [inline src/threeIdProvider.ts:41]: We should probably replace these by the error codes and messages defined in https://eips.ethereum.org/EIPS/eip-1193 so better consistency

File: src/threeIdProvider.ts [hunk 3]
```diff
@@ -19,94 +20,98 @@ type Context = {
...
-
-    wallet.events.on('new-auth-method', (authBlob: EncryptedMessage) => {
-      this._newAuthMethods.push(authBlob)
-    })
-    wallet.events.on('new-link-proof', (linkProof: LinkProof) => {
-      this._newLinks.push(linkProof)
-    })
+  constructor({ permissions, threeIdx, keyring, forcedOrigin }: ProviderConfig) {
+    this._handle = (origin: string, req: RPCRequest) => {
+      return createHandler<Context>(Object.assign(methods, didMethods))({
```
Comment (PaulLeCam) [inline src/threeIdProvider.ts:100]: As commented above, it's not ideal to call `createHandler` for every call to handle

File: src/utils.ts [hunk 4]
```diff
@@ -17,6 +19,39 @@ export const sha256Multihash = (str: string): string => {
...
+  }
+  buf[0] = multicodecPubkeyTable[keyType]
+  // The multicodec is encoded as a varint so we need to add this.
+  // See js-multicodec for a general implementation
+  buf[1] = 0x01
+  buf.set(key, 2)
+  return `z${bs58.encode(buf)}`
+}
+
+export const gen3IDgenesis = (pubkeys: any): any => {
```
Comment (PaulLeCam) [inline src/utils.ts:41]: Maybe we could allow extra keys to be provided here so we could add the `idx` field?

Comment (oed) [reply src/utils.ts:41]: `idx` can not be part of the genesis record. Will add the idx property and documents in the upcoming PR :)

File: src/did-provider.ts [hunk 5]
```diff
@@ -9,52 +9,78 @@ import {
...
 export class DidProvider implements RPCConnection {
   protected _handle: RequestHandler
   protected _wallet: IdentityWallet
 
-  constructor(wallet: IdentityWallet) {
-    this._handle = createHandler<Context>(methods)
-    this._wallet = wallet
+  constructor({ permissions, threeIdx, keyring, forcedOrigin }: ProviderConfig) {
+    this._handle = (origin: string, msg: RPCRequest) => {
+      return createHandler<Context>(didMethods)({
```
[the remark this answers is not in the document]
Comment (oed) [reply src/did-provider.ts:70]: Thanks, this is an easy fix here anyhow :+1:

File: src/permissions.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  async request(origin: string, paths: Array<string> = []): Array<string> | null {
+    if (this.has(origin, paths)) {
+      return paths
+    } else {
+      const given = await this.getPermission({
+        type: 'authenticate',
+        origin,
+        payload: { paths }
+      })
+      this.set(origin, given)
```
[the remark this answers is not in the document]
Comment (oed) [reply src/permissions.ts:42]: What would you recommend doing here instead?

[the remark this answers is not in the document]
Comment (PaulLeCam) [reply src/permissions.ts:42]: I think the solution would be to merge the `given` paths with the existing ones in the `set` method, so the result of multiple calls would still be deterministic. It should work well as long as the merging logic is simple, but it still relies on the `getPermission` logic not to mess up its prompts.
Another solution would be to chain the `getPermission` calls to make sure there can be only one in flight, this way even a naive implementation of `getPermission` should provide the expected UX.

[the remark this answers is not in the document]
Comment (oed) [reply src/permissions.ts:50]: Did only the merging for now. Should be fine for the time being.

File: src/threeIdProvider.ts [hunk 7]
```diff
@@ -8,105 +8,111 @@ import {
...
+          uncompressed: true,
+        })
+        return acc
+      }, {} as Record<string, PublicKeys>),
+    }
   },
-  '3id_isAuthenticated': async (ctx, params) => {
-    return await ctx.provider.wallet.isAuthenticated(params.spaces, ctx.origin)
+  '3id_isAuthenticated': async ({ permissions, origin }, params) => {
+    return Promise.resolve(permissions.has(origin, params.spaces))
```
Comment (PaulLeCam) [inline src/threeIdProvider.ts:33]: Not an issue but it seems this function doesn't have to be `async` / using `Promise.resolve`

Comment (oed) [reply src/threeIdProvider.ts:33]: Doesn't it have to since the `send` method awaits the `_handle` method?

Comment (PaulLeCam) [reply src/threeIdProvider.ts:33]: AFAIK the way `await` is implemented it checks if the result is a promise or not to extract the value, so it works either way, but might be safer not to rely on this behavior anyways.

**Pull request merged**